### ISSUE TITLE
Allow multiple providers to access the same configuration values

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi#12930]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently, the namespaces in configuration are tied directly to the provider's name (e.g. `aws` or `aws-native`). The engine only passes a subset of the Pulumi configuration to each provider - filtered by its namespace.

It would be beneficial for a provider to be able to request a different, or possibly multiple namespaces to retrieve its configuration from.

### Use Case 1 - different providers for the same cloud

Both the `aws` and `aws-native` providers should have identical configuration. For customers using both providers together, they have duplication of all their provider settings. This could be avoided if the aws-native provider could first check its own `aws-native` namespace, but also access the settings from the `aws` namespace as a fallback. Similarly, the `aws` provider could also check for `aws-classic` and `aws` keys.

This would mean the settings namespace could refer to the cloud rather than just the provider.

### Use Case 2 - breaking up large providers

The `azure-native` provider and SDKs are very big. We would like the option to ship them in smaller parts. One option would be to ship the main provider with a hand-picked set of resources across all services, then ship a separate provider for each Azure service - which then includes every version of every resource. A blocker to doing this is that the user would have to re-specify the same configuration for each separate provider they used.

**Comment by justinvp** [comment]
Any ideas on how we'd mitigate a "bad" provider requesting config of another provider and then gaining access to credentials that it shouldn't have?

**Comment by Frassle** [comment]
> Any ideas on how we'd mitigate a "bad" provider requesting config of another provider and then gaining access to credentials that it shouldn't have?

I wonder how much we need to actually care about that. Pretty sure we'll send all config to component providers anyway as part Call/Construct. And if someones running a bad provider there's nothing stopping it from reading the Pulumi.stack.yaml file and looking at envvars itself as just a local process.

**Comment by lblackstone** [comment]
> > Any ideas on how we'd mitigate a "bad" provider requesting config of another provider and then gaining access to credentials that it shouldn't have?
> 
> I wonder how much we need to actually care about that. Pretty sure we'll send all config to component providers anyway as part Call/Construct. And if someones running a bad provider there's nothing stopping it from reading the Pulumi.stack.yaml file and looking at envvars itself as just a local process.

Agreed. Plus, the user program can also request all of that config, so I'm not sure this feature would materially change the security posture.

**Comment by AaronFriel** [comment]
@pgavlin and I have had a few long chats about the value of keeping plugins remote-able and he's made me a convert. It keeps us open to doing things in the future like run plugins in sandboxed processes, containers, or remotely.

Just because plugins currently all run on the same machine does not mean they always will, and making this explicit (via an RPC response in Configure() & schema?) will help us debug issues and help incident response teams in the future.

**Comment by AaronFriel** [comment]
My response is to the security posture comments, because the current providers run locally and therefore could likely already do this via "trickery" (reading parent process info, whatever). I'm saying that wasn't the intent of the plugin architecture, and we'd be basing a security posture decision on what is effectively a coincidence or convenience that is true today.

If we commit to all providers getting all config, it's a breaking change to change that when we do want to implement isolation. Given our breaking change policy, I really don't want to commit to that.

I don't like the idea that a provider could - today, we don't know if any do - log the credentials they receive or send them upstream. The blast radius right now of that is just that provider's config. It's very easy to accidentally store more than you intended:

https://krebsonsecurity.com/2019/03/facebook-stored-hundreds-of-millions-of-user-passwords-in-plain-text-for-years/

So I think we should be explicit about this, and make this a feature providers opt into, or even special case a few providers in the engine for this purpose.

**Comment by lblackstone** [comment]
> If we commit to all providers getting all config, it's a breaking change to change that when we do want to implement isolation. Given our breaking change policy, I really don't want to commit to that.

I think that ship has sailed already with https://github.com/pulumi/pulumi/pull/12605

**Comment by justinvp** [comment]
> I think that ship has sailed already with #12605

With #12605 you only get the config for your provider. You don't get any config associated with other providers.

**Comment by lblackstone** [comment]
> > I think that ship has sailed already with #12605
> 
> With #12605 you only get the config for your provider. You don't get any config associated with other providers.

Oh, right. I forgot that's what we landed on. There was a similar discussion on that one.

**Comment by danielrbradley** [comment]
> If we commit to all providers getting all config

That's not what I'm proposing - I'm suggesting providers can request specific config namespaces - not all config.

Perhaps there's other approaches we could use here to share config between providers at a config level rather than at the point of the provider requesting it? 

Off the top of my head, you could solve the first use case conceptually like this (yeah I know our config isn't actually nested like this though):

```yaml
config:
  aws:
    region: eu-west-1
  aws-native:
    extends: aws:*
```

**Comment by AaronFriel** [comment]
Ah, my mistake! I understood the replies above to be about why it would be OK to send all config.

As long as this is explicit - and ideally documented in schema so that eventually it could be enforced by the engine - then I am very on board.

**Comment by danielrbradley** [comment]
@t0yv0 and @lblackstone how does this align with the shared configuration work you were looking at? Is this the same problem or a slightly different variation?

**Comment by t0yv0** [comment]
Not sure anything I looked at addresses this. 

Besides security I'd also think about debug-ability and the config systems being tractable. Implicit flow of information can be confusing. I would have a slight preference that this flow of information is made explicit in the configuration language and opt-in. Either shared expressions or patterns. 

Patterns:

```
config:
  aws*:region: us-west-2
  azure*:location: westus2
```

Expressions:

```
shared:
  azloc: westus2
config:
  azure-native-1:location: ${azloc}
  azure-native-2:location: ${azloc}
  azure-native-3:location: ${azloc}
```

Feature-rich config languages like Cue and DHall provide a more consistent way of sharing expressions on to reduce redundancy in YAML, for something more thought-out and battle-tested.

**Comment by lblackstone** [comment]
> how does this align with the shared configuration work you were looking at?

This is one of the options that we discussed. I think that it would be a useful feature to support.
